source/develop/release-management/features/virt/dedicated-cpus.md

```diff
@@ -0,0 +1,68 @@
+---
+Title: Dedicated CPUs
+Category: feature
+Authors: sabusale
+---
+
+# Dedicated CPUs
+
+## Summary
+
+Once choosing dedicated CPU policy, The guest vCPUs will be strictly pinned to a set of host pCPUs (similarly to static CPU pinning). The set of pCPUs will be chosen to match the required guest CPU topology.
+
+## Owner
+
+*   Name: Saif Abu Saleh (sabusale)
+*   Email: sabusale@redhat.com
+
+
+### Tasks:
+
+
+1- Add dedicated to CPU Pinning Policy enum
+
+2- Scheduling
+
+3- Send vcpupin with cpusets and policy name in VM metadata(only when Policy is dedicated?)
+
+4- Send the policy name from vmStatic policy field
+
+5- Send the cpusets (convert from CurrentCpuPinningField form to array)
+
+6- API: Migrate
+Send cpusets additional parameter (convert from CurrentCpuPinningField form to array)
+
+7- setNumberOfCpus VDSM call
+Send cpusets additional parameter (convert from CurrentCpuPinningField form to array)
+Frontend: Add Dedicated policy dropdown to the UI
+
+
+Steps 3-7: can be done in parallel by different person
+
+
+### Scheduling:
+
+* Find host that have the VM CPU topology on the scheduling phase
+* On run phase, schedule the VM to run on that host
+
+
+###Data Model:
+
+* Save CPU mappings in a field (maybe use CurrentCpuPinningField)?
+* Add used_cpus field to the host to know how many CPUs is used
```

Comment:
To make it easier, used_cpus can be an array holding the pCPUs taken(to prevent later calculating of the cpu pinning string). Then you can filter easily by checking its length, while you can easily choose pCPU for your VM.

Reply:
how it will make it easier?

I need mapping from vCPU to pCPU per VM anyway (to send the cpusets)

You mean I need it only to know which pCPU is currently available right?

Reply:
I think we should not keep the "used" CPUs but the "available" - as is done in the vdsm. In fact, I think the vdsm should report those available CPUs (called shared_pool I think) and we should work with that, not to keep our own copy of the data (except some kind of "pending" cpus as is used in scheduler even now)